File: api-gateway/core/grpc_client.py
import grpc.aio
import os
import library_service_pb2_grpc

# Async gRPC client connection
async def get_grpc_client():
    grpc_host = os.getenv('GRPC_SERVER_HOST', 'localhost')
    grpc_port = os.getenv('GRPC_SERVER_PORT', '50051')
    channel = grpc.aio.insecure_channel(f'{grpc_host}:{grpc_port}')
    return library_service_pb2_grpc.LibraryServiceStub(channel)

File: api-gateway-node/utils/grpcClient.js
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const path = require('path');

// Load protobuf
const PROTO_PATH = path.join(__dirname, '../proto/library_service.proto');
const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true
});

const libraryProto = grpc.loadPackageDefinition(packageDefinition).library;

// gRPC client
const grpcHost = process.env.GRPC_SERVER_HOST || 'localhost';
const grpcPort = process.env.GRPC_SERVER_PORT || '50051';
const client = new libraryProto.LibraryService(`${grpcHost}:${grpcPort}`, grpc.credentials.createInsecure());

// Helper function to promisify gRPC calls
const grpcCall = (method, request) => {
  return new Promise((resolve, reject) => {
    client[method](request, (error, response) => {
      if (error) reject(error);
      else resolve(response);
    });
  });
};

module.exports = { grpcCall };